src/python/gem5/utils/multisim/multisim.py

```diff
@@ -0,0 +1,299 @@
+    # module path (the config script specifying all simulations using MultiSim)
+    # but a different ID. The ID is used to select the correct simulator to
+    # run.
+    pool.starmap(_run, zip([module_path for _ in range(len(ids))], tuple(ids)))
+
+
+def set_num_processes(num_processes: int) -> None:
+    """Set the max number of processes to run in parallel.
+
+    :param num_processes: The number of processes to run in parallel.
+    """
+    if num_processes < 1:
+        raise ValueError("Number of processes must be greater than 0.")
+    if isinstance(num_processes, int):
+        global _num_processes
+        _num_processes = num_processes
+    else:
+        raise ValueError("Number of processes must be an integer.")
+
+
+def num_simulators() -> int:
+    """Returns the number of simulators added to the MultiSim."""
+    return len(_multi_sim)
+
+
+def add_simulator(simulator: "Simulator") -> None:
+    """Add a single simulator to the Multisim. Doing so informs the simulators
+    to run this simulator via multiprocessing.
+
+    **Note:** If this function is not run using the MultiSim module then the
+    user will be prompted to either do so if they desire multiple gem5
+    processes or to pass the id of the simulator to run. This function will
+    attempt to run the simulation with the id passed as an argument. If
+    such simulation exists the simulation will end without failure (or any
+    simulations having been run).
+
+    :param simulator: The simulator to add to the multisim.
+    :param id: The id of the simulator. This is used to reference the
+    simululation. This is particularly important when referencing the correct
+    m5out subdirectory.
+    """
+
+    global _multi_sim
+    if not simulator.get_id():
+        # The default simulator id is the length of the current set of
+        # simulators. This is used to ensure that the simulator has a unique
+        # id.
+        simulator.set_id(f"sim_{len(_multi_sim)}")
+    _multi_sim.add(simulator)
+
+    # The following code is used to enable a user to run a single simulation
+    # from the config script, based on an ID, in the case the config script is
+    # passed a traditional gem5 config and not via the multisim module.
+    global module_run
+    if not module_run:
+        import argparse
+
+        parser = argparse.ArgumentParser(
+            description="Run a specific simulation based on the id."
+        )
+        parser.add_argument(
+            "id",
+            type=str,
+            nargs="?",
+            default=None,
+            help="The id of the simulator to run.",
+        )
+        parser.add_argument(
+            "-l",
+            "--list",
+            help="List the ids of the simulators to run.",
+            action="store_true",
+        )
+        args = parser.parse_args()
+        if args.list:
+            print(simulator.get_id())
+        elif not args.id:
```

Comment:
If you specify an id, but it doesn't match, nothing happens. I think we should explicitly have an error in that case.

Reply:
This may be hard so I'm gonna drop it for now. Warning to users: Without specifying a valid ID this will finish without running anything and without error.